File: frontend/src/api/chapters.js
import client from './client'

export const chaptersApi = {
  list: (projectId) => client.get(`/projects/${projectId}/chapters`).then((r) => r.data),
  get: (chapterId) => client.get(`/chapters/${chapterId}`).then((r) => r.data),
  create: (projectId, payload) =>
    client.post(`/projects/${projectId}/chapters`, payload).then((r) => r.data),
  update: (chapterId, payload) =>
    client.patch(`/chapters/${chapterId}`, payload).then((r) => r.data),
  remove: (chapterId) => client.delete(`/chapters/${chapterId}`),
  reorder: (projectId, chapterIds) =>
    client
      .post(`/projects/${projectId}/chapters/reorder`, { chapter_ids: chapterIds })
      .then((r) => r.data),
  saveContent: (chapterId, content) =>
    client.put(`/chapters/${chapterId}/content`, { content }).then((r) => r.data),
  // AI 草拟章节节拍。返回 {beats: [...]},不直接落库。
  suggestBeats: (chapterId, payload = {}) =>
    client
      .post(`/chapters/${chapterId}/ai/suggest-beats`, payload, { timeout: 90000 })
      .then((r) => r.data),
  // 单章自动索引:AI 写完正文落地后调,把这章的 plot_events 全删重抽,
  // 避免下次生成漏掉刚发生的事(沉默失败)。
  autoIndex: (chapterId) =>
    client
      .post(`/chapters/${chapterId}/auto-index`, {}, { timeout: 90000 })
      .then((r) => r.data),
  // 节拍-事件对账:写完 + 索引后调,逐拍判断 covered/partial/missing,
  // 结果落 chapter.beats_alignment。
  checkBeats: (chapterId) =>
    client
      .post(`/chapters/${chapterId}/ai/check-beats`, {}, { timeout: 90000 })
      .then((r) => r.data),
}

File: backend/app/api/chapters.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.schemas.chapter import (
    ChapterContentSaved,
    ChapterContentUpdate,
    ChapterCreate,
    ChapterDetail,
    ChapterListItem,
    ChapterReorder,
    ChapterUpdate,
)
from app.services import chapter_service
from app.services.chapter_service import (
    ChapterNotFoundError,
    ProjectNotFoundForChapterError,
    ReorderMismatchError,
)

# 工程下的章节集合
project_router = APIRouter(prefix="/api/projects/{project_id}/chapters", tags=["chapters"])
# 单个章节
chapter_router = APIRouter(prefix="/api/chapters", tags=["chapters"])


@project_router.get("", response_model=list[ChapterListItem])
def list_chapters(project_id: int, db: Session = Depends(get_db)) -> list[ChapterListItem]:
    try:
        return chapter_service.list_chapters(db, project_id)
    except ProjectNotFoundForChapterError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="工程不存在") from e


@project_router.post("", response_model=ChapterDetail, status_code=status.HTTP_201_CREATED)
def create_chapter(
    project_id: int, payload: ChapterCreate, db: Session = Depends(get_db)
) -> ChapterDetail:
    try:
        return chapter_service.create_chapter(db, project_id, payload)
    except ProjectNotFoundForChapterError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="工程不存在") from e


@project_router.post("/reorder", response_model=list[ChapterListItem])
def reorder_chapters(
    project_id: int, payload: ChapterReorder, db: Session = Depends(get_db)
) -> list[ChapterListItem]:
    try:
        return chapter_service.reorder_chapters(db, project_id, payload.chapter_ids)
    except ProjectNotFoundForChapterError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="工程不存在") from e
    except ReorderMismatchError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)) from e


@chapter_router.get("/{chapter_id}", response_model=ChapterDetail)
def get_chapter(chapter_id: int, db: Session = Depends(get_db)) -> ChapterDetail:
    try:
        return chapter_service.get_chapter(db, chapter_id)
    except ChapterNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="章节不存在") from e


@chapter_router.patch("/{chapter_id}", response_model=ChapterDetail)
def update_chapter(
    chapter_id: int, payload: ChapterUpdate, db: Session = Depends(get_db)
) -> ChapterDetail:
    try:
        return chapter_service.update_chapter(db, chapter_id, payload)
    except ChapterNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="章节不存在") from e


@chapter_router.delete("/{chapter_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_chapter(chapter_id: int, db: Session = Depends(get_db)) -> None:
    try:
        chapter_service.delete_chapter(db, chapter_id)
    except ChapterNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="章节不存在") from e


@chapter_router.put("/{chapter_id}/content", response_model=ChapterContentSaved)
def save_chapter_content(
    chapter_id: int, payload: ChapterContentUpdate, db: Session = Depends(get_db)
) -> ChapterContentSaved:
    try:
        c = chapter_service.save_content(db, chapter_id, payload.content)
    except ChapterNotFoundError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="章节不存在") from e
    return ChapterContentSaved.model_validate(c)
